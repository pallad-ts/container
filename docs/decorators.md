# Decorators
Creating service definition via container (`definitionWithConstructor` etc) might be a tedious and error-prone work.
That's why _alpha-dic_ supports decorators for classes.

NOTE! Hopefully this is obvious but Decorators are not supported by node.js natively. You need to use Typescript (highly recommended) or other transpiler (like Babel) to make it work.

```typescript
const {Service, Annotation, Inject} = require('alpha-dic');

@Service() // you can also provide a name @Service('name'). By default class/function name inferred.
@Annotation({name: 'extraAnnotation'})
export class NotificationCenter {
    
    @Inject('superExtraService')
    public extraService: ExtraService;
    
    constructor(
        @Inject('EmailRenderer') emailRenderer: EmailRenderer, 
        @Inject('EmailSender') emailSender: EmailSender) {
    }
}

```

## Setup

Since decorator is just regular function you need to give it an opportunity to get called. To do soo you can use `loadServices` function that loads all modules matching given patterns.  
```typescript
import {createStandard, loadServices} from 'alpha-dic';

const container = createStandard();

// loads all files (with extensions .ts, .js, .tsx but ignores .d.ts, .d.tsx) in given directory so all @Service decorator calls got a chance to register service definitions
loadServices(container, {
    patterns: './services/*.*' // note ".*" to make sure glob can catch all files
}); 
```