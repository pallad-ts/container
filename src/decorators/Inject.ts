import {ContainerArg} from "../ContainerArg";
import 'reflect-metadata';
import {Reference} from "../Reference";
import * as is from 'predicates';
import {ensureMetadata} from "./serviceMetadata";


const assertServiceNameOrContainerArg = is.assert(
    is.any(is.string, is.instanceOf(ContainerArg)),
    '@Inject argument must be a string that represents service name or an object of ContainerArg instance'
);

export function Inject(serviceName: string | ContainerArg) {
    assertServiceNameOrContainerArg(serviceName);

    const arg = is.string(serviceName) ? Reference.one.name(serviceName) : serviceName;
    return function (target: any, property: string | symbol, indexOrDescriptor?: number | TypedPropertyDescriptor<any>) {
        const isParameterDecorator = typeof indexOrDescriptor === 'number';
        if (isParameterDecorator) {
            // argument decorator
            ensureMetadata(target).constructorArguments[<number>indexOrDescriptor] = arg;
        } else {
            // property decorator
            ensureMetadata(target.constructor).propertiesInjectors.set(property, arg);
        }
    }
}