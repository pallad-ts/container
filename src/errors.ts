import {create, DomainErrorDescriptor, generators} from 'alpha-errors';

const domain = create({
    codeGenerator: generators.formatCode('ADIC_%d')
});

export const SERVICE_NOT_FOUND = domain.create('Service not found');
export const AMBIGUOUS_SERVICE = domain.create('Ambiguous service');
export const NO_MATCHING_SERVICE = domain.create('No matching service');
export const CIRCULAR_DEPENDENCY_DETECTED = domain.create('Circular dependency detected');
export const INCOMPLETE_DEFINITION = domain.create('The service definition is not complete');
export const NO_SERVICE_NAME = domain.create('Missing service name');
export const INVALID_SERVICE_ARGUMENTS_LENGTH = domain.create('Invalid service arguments length');
export const MISSING_INJECT_DECORATOR = domain.create('Missing @Inject decorator');
export const MISSING_CONFIG_VALUE = domain.create('Missing config value');