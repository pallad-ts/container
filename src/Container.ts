import {
    AnnotationPredicate, ServiceName, DefinitionPredicate, Middleware,
    ServiceFactory, onMiddlewareAttach
} from './types';
import {Definition} from './Definition';
import * as errors from './errors';
import {assertNoCircularDependencies} from './assertNoCircularDependencies';
import {ContainerArg} from './args/ContainerArg';
import * as is from 'predicates';
import {randomName} from './randomName';
import {debugFn} from './debugFn';

const debugCreation = debugFn('creation');
const debugDefinition = debugFn('definition');
const debugSlowCreation = debugFn('slow-creation');

export class Container {
    private definitions: Map<string | Symbol, Definition> = new Map();
    private services: Map<Definition, Promise<any>> = new Map();
    private middlewares: Middleware[] = [];

    public readonly parent?: Container;

    /**
     * Time needed to trigger debug slow creation log
     */
    public slowLogThreshold: number = 10000;

    constructor(parent?: Container) {
        Object.defineProperty(this, 'parent', {
            value: parent,
            writable: false
        });
    }

    /**
     * Registers given service definition
     */
    registerDefinition(definition: Definition): this {
        if (this.definitions.has(definition.name)) {
            throw errors.ALREADY_DEFINED(`Service "${definition.name.toString()}" already defined`);
        }
        debugDefinition(`Service ${definition.name.toString()} defined`);
        definition.setOwner(this);
        this.definitions.set(definition.name, definition);
        return this;
    }

    /**
     * Creates and registers service definition
     *
     * Returns created definition for further configuration
     */
    definition(name?: ServiceName) {
        const definition = new Definition(name);
        this.registerDefinition(definition);
        return definition;
    }

    /**
     * Creates and registers service definition with given name, function as constructor
     */
    definitionWithConstructor(name: ServiceName, clazz: { new(...args: any[]): any }): Definition;
    definitionWithConstructor(clazz: { new(...args: any[]): any }): Definition;
    definitionWithConstructor(nameOrClazz: ServiceName | { new(...args: any[]): any }, clazz?: { new(...args: any[]): any }) {
        const finalClazz = is.func(nameOrClazz) ? nameOrClazz : clazz!;
        return this.definition(ServiceName.is(nameOrClazz) ? nameOrClazz : randomName(finalClazz.name))
            .useConstructor(finalClazz);
    }

    /**
     * Creates and registers service definition with given name, function as factory
     */
    definitionWithFactory(name: ServiceName, factory: ServiceFactory, type?: Function): Definition;
    definitionWithFactory(factory: ServiceFactory, type?: Function): Definition;
    definitionWithFactory(nameOrFactory: ServiceName | ServiceFactory, factoryOrType: ServiceFactory | Function, type?: Function): Definition {
        const name = ServiceName.is(nameOrFactory) ? nameOrFactory : undefined;
        const factory = is.func(nameOrFactory) ? nameOrFactory as ServiceFactory : factoryOrType as ServiceFactory;
        const finalType = is.func(nameOrFactory) ? factoryOrType : type;

        const def = this.definition(name)
            .useFactory(factory);
        if (finalType) {
            def.markType(finalType);
        }
        return def;
    }

    /**
     * Creates and registers service definition with given name and value as a service
     */
    definitionWithValue(name: ServiceName, value: any): Definition;
    definitionWithValue(value: any): Definition;
    definitionWithValue(nameOrValue: ServiceName | any, value?: any): Definition {
        const isNameProvided = value !== undefined && ServiceName.is(nameOrValue);
        const finalValue = isNameProvided ? value : nameOrValue;
        const name = isNameProvided ? nameOrValue : randomName(Object.getPrototypeOf(finalValue).constructor.name);
        return this.definition(name)
            .useValue(finalValue);
    }

    /**
     * Returns definition by given name
     */
    findByName(name: ServiceName): Definition | undefined {
        if (this.definitions.has(name)) {
            return this.definitions.get(name);
        }

        if (this.parent) {
            return this.parent.findByName(name);
        }
    }

    /**
     * Returns definitions that satisfy given predicate
     */
    findByPredicate(predicate: DefinitionPredicate): Definition[] {
        return Array.from(this.definitions.values())
            .filter(predicate)
            .concat(this.parent ? this.parent.findByPredicate(predicate) : []);
    }


    /**
     * Returns all definitions that contain annotation that satisfied given predicate
     */
    findByAnnotation(predicate: AnnotationPredicate): Definition[];
    findByAnnotation(predicate: AnnotationPredicate, withAnnotation: false): Definition[];
    findByAnnotation(predicate: AnnotationPredicate, withAnnotation: true): Array<[Definition, any]>;
    findByAnnotation(predicate: AnnotationPredicate, withAnnotation: boolean = false): Definition[] | Array<[Definition, any]> {
        let annotations: any[] = [];
        const definitions = this.findByPredicate(s => {
            const annotation = s.annotations.find(predicate);

            if (annotation) {
                withAnnotation && annotations.push(annotation);
                return true;
            }
            return false;
        });

        if (withAnnotation) {
            return definitions.map((d): [Definition, any] => {
                return [d, annotations.shift()];
            });
        }
        return definitions;
    }

    /**
     * Registers given middleware
     */
    addMiddleware(...middlewares: Middleware[]): Container {
        for (const middleware of middlewares) {
            if (middleware[onMiddlewareAttach]) {
                middleware[onMiddlewareAttach]!(this);
            }
        }
        this.middlewares.push(...middlewares);
        return this;
    }

    getMiddlewares(): Middleware[] {
        return this.middlewares;
    }

    /**
     * Returns service for given name or definition
     */
    get<T = any>(nameOrDefinition: ServiceName | Definition): Promise<T> {
        let definition: Definition | undefined;

        if (ServiceName.is(nameOrDefinition)) {
            definition = this.findByName(nameOrDefinition);
            if (definition === undefined) {
                return Promise.reject(errors.SERVICE_NOT_FOUND(`Service "${nameOrDefinition.toString()}" does not exist`));
            }
        } else {
            definition = nameOrDefinition;
        }

        if (this.services.has(definition)) {
            return this.services.get(definition)!;
        }

        if (!this.definitions.has(definition.name) && this.parent) {
            return this.parent.get(definition);
        }

        const promise = this.create(definition);
        this.services.set(definition, promise);
        return promise;
    }

    private async create(definition: Definition) {
        if (!definition.factory) {
            return Promise.reject(
                errors.INCOMPLETE_DEFINITION(
                    `Missing factory for service definition "${definition.name.toString()}". Define it as constructor, factory or value`
                )
            );
        }

        assertNoCircularDependencies(this, definition);

        // valid definition, time to lock it
        definition.lock();

        let timeout: any;
        if (this.slowLogThreshold > 0) {
            timeout = setTimeout(() => {
                debugSlowCreation(`Service ${definition.name.toString()} takes a long time to create. Over ${this.slowLogThreshold} ms`);
            }, this.slowLogThreshold);
        }

        const debugMsg = `Creating service ${definition.name.toString()}`;
        debugCreation(`${debugMsg} - started`);
        let currentMiddleware = 0;
        const middlewares = this.getMiddlewares();
        const next = (definition: Definition) => {
            const middleware = middlewares[currentMiddleware];
            currentMiddleware++;
            if (middleware) {
                return middleware.call(this, definition, next);
            }

            return Promise.all(
                definition.args.map(a => ContainerArg.is(a) ? a.getArgument(this) : a)
            )
                .then((args: any[]) => definition.factory.apply(this, args));
        };

        try {
            const result = await next.call(this, definition);
            timeout && clearTimeout(timeout);
            debugCreation(`${debugMsg} - finished`);
            return result;
        } catch (e) {
            timeout && clearTimeout(timeout);
            throw e;
        }
    }

    /**
     * Returns all services that definition satisfies predicate
     */
    getByPredicate<T = any>(predicate: DefinitionPredicate): Promise<T[]> {
        return Promise.all(
            this.findByPredicate(predicate)
                .map(d => this.get(d))
        );
    }

    /**
     * Returns all services that definition contains annotation that satisfies given predicate
     */
    getByAnnotation<T = any>(predicate: AnnotationPredicate): Promise<T[]>;
    getByAnnotation<T = any>(predicate: AnnotationPredicate, withAnnotation: false): Promise<T[]>;
    getByAnnotation<T = any>(predicate: AnnotationPredicate, withAnnotation: true): Promise<Array<[T, any]>>;
    getByAnnotation<T = any>(predicate: AnnotationPredicate, withAnnotation: boolean = false): Promise<T[] | Array<[T, any]>> {
        return Promise.all(
            this.findByAnnotation(predicate, true)
                .map(async ([definition, annotation]) => {
                    if (withAnnotation) {
                        return [await this.get(definition), annotation];
                    }
                    return this.get(definition);
                })
        );
    }

    alias(definition: Definition, options?: {
        name?: string,
        withAnnotations?: boolean | ((ann: any) => boolean)
    }) {
        const def = this.definitionWithFactory(options?.name ? options.name : definition.name, () => {
            if (!definition.owner) {
                throw errors.DEFINITION_WITHOUT_CONTAINER(
                    `Cannot create service "${def.name.toString()}" due to lack of assigned container`
                );
            }
            return definition.owner.get(definition);
        });

        const annotations = (() => {
            const withAnnotations = options?.withAnnotations ?? false;
            if (is.func(withAnnotations)) {
                return definition.annotations.filter(withAnnotations);
            }
            return withAnnotations ? definition.annotations : [];
        })();

        if (annotations.length > 0) {
            def.annotate(...annotations);
        }
        if (definition.type) {
            def.markType(definition.type);
        }
        return def;
    }
}