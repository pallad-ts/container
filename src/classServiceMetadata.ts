import { Definition } from "./Definition";
import { fromConstructor } from "./serviceFactories";
import { ClassConstructor, ServiceName } from "./types";
import { ERRORS } from "./errors";

require("reflect-metadata");

const METADATA_KEY = Symbol("@pallad/container/ServiceMetadata");
const DEFINITION_KEY = Symbol("@pallad/container/Definition");

export interface ClassServiceMetadata {
	name?: ServiceName;
	clazz: ClassConstructor<any>;
	constructorArguments: any[];
	propertiesInjectors: Map<string | symbol, any>;
	annotations: any[];
}

/**
 * Creates service metadata data for given class if it doesn't exist
 */
export function ensureMetadataAttachedToClass(clazz: ClassConstructor<any>): ClassServiceMetadata {
	let data: ClassServiceMetadata = Reflect.getMetadata(METADATA_KEY, clazz);
	if (!data) {
		data = {
			clazz: clazz,
			propertiesInjectors: new Map(),
			constructorArguments: [],
			annotations: [],
		};
		Reflect.defineMetadata(METADATA_KEY, data, clazz);
	}
	return data;
}

export function getClassServiceMetadata(clazz: ClassConstructor<any>): ClassServiceMetadata | undefined {
	return Reflect.getMetadata(METADATA_KEY, clazz);
}

export function extractDefinitionFromClass<T>(clazz: ClassConstructor<T>) {
	const definition = Reflect.getMetadata(DEFINITION_KEY, clazz);
	if (definition) {
		return definition;
	}
	let data: ClassServiceMetadata = Reflect.getMetadata(METADATA_KEY, clazz);
	if (data) {
		const definition = createDefinitionFromMetadata(data, clazz);
		Reflect.defineMetadata(DEFINITION_KEY, definition, clazz);
		return definition;
	}

	throw ERRORS.CLASS_IS_NOT_A_SERVICE.create(clazz.name);
}

function createDefinitionFromMetadata(
	metadata: ClassServiceMetadata,
	constructor: ClassConstructor<any>
) {
	assertValidServiceDefinition(constructor, metadata);
	const definition = Definition.useFactory(
		(...args: unknown[]) => {
			const constructorArgs = args.slice(0, metadata.constructorArguments.length);
			const service = fromConstructor(constructor)(...constructorArgs);

			// inject to properties
			const propertiesInjectionsStartIndex = metadata.constructorArguments.length;
			let propertyInjectionIndex = 0;
			for (const [propertyName] of metadata.propertiesInjectors.entries()) {
				service[propertyName] =
					args[propertiesInjectionsStartIndex + propertyInjectionIndex++];
			}
			return service;
		},
		{
			name: metadata.name,
			type: constructor,
		}
	);

	const args = metadata.constructorArguments.concat(
		Array.from(metadata.propertiesInjectors.values())
	);

	return definition.withArguments(...args).annotate(...metadata.annotations);
}

function assertValidServiceDefinition(constructor: Function, metadata: ClassServiceMetadata) {
	if (constructor.length > metadata.constructorArguments.length) {
		throw ERRORS.INVALID_SERVICE_ARGUMENTS_LENGTH.create(
			metadata.name!.toString(),
			constructor.length,
			metadata.constructorArguments.length
		);
	}

	for (let i = 0; i < metadata.constructorArguments.length; i++) {
		const arg = metadata.constructorArguments[i];
		if (!arg) {
			throw ERRORS.MISSING_INJECT_DECORATOR.create(i);
		}
	}
}
