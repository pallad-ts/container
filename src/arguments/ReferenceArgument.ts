import { AnnotationPredicate, ServiceName, DefinitionPredicate } from "../types";
import { Container } from "../Container";
import { Definition } from "../Definition";
import { Lookup } from "../Lookup";
import { ContainerArgument } from "./ContainerArgument";
import { TypeReference } from "../TypeReference";
import { ERRORS } from "../errors";

function toTypeRef(type: TypeReference | Function) {
	return TypeReference.is(type) ? type : new TypeReference(type);
}

export class ReferenceArgument extends ContainerArgument<unknown> {
	static one = {
		name(name: ServiceName) {
			return new ReferenceArgument("one", new Lookup.ByServiceName(name));
		},
		predicate(predicate: DefinitionPredicate) {
			return new ReferenceArgument("one", new Lookup.ByPredicate(predicate));
		},
		annotation<T>(predicate: AnnotationPredicate<T>) {
			return new ReferenceArgument("one", new Lookup.ByAnnotation(predicate));
		},
		type(type: TypeReference | Function) {
			return new ReferenceArgument("one", new Lookup.ByType(toTypeRef(type)));
		},
	};

	static multi = {
		predicate(predicate: DefinitionPredicate) {
			return new ReferenceArgument("multi", new Lookup.ByPredicate(predicate));
		},
		annotation(predicate: AnnotationPredicate<unknown>) {
			return new ReferenceArgument("multi", new Lookup.ByAnnotation(predicate));
		},
		type(type: TypeReference | Function) {
			return new ReferenceArgument("multi", new Lookup.ByType(toTypeRef(type)));
		},
	};

	constructor(
		private readonly type: "one" | "multi",
		private readonly lookup: Lookup
	) {
		super();
		Object.freeze(this);
	}

	resolve(container: Container): Promise<any> {
		const definitions = Array.from(this.findDefinitions(container));

		if (this.type === "one") {
			const definition = definitions[0];

			return container.resolve(definition);
		}

		return Promise.all(
			definitions.map(x => {
				return container.resolve(x);
			})
		);
	}

	*dependencies(container: Container) {
		yield* this.findDefinitions(container);
	}

	private *findDefinitions(container: Container): Generator<Definition> {
		if (this.type === "one") {
			yield this.findOne(container);
		} else {
			yield* this.findMulti(container);
		}
	}

	private findOne(container: Container): Definition {
		const definitions = Array.from(this.lookup.find(container));
		if (definitions.length === 0) {
			throw ERRORS.NO_MATCHING_SERVICE.create(this.lookup);
		}

		if (definitions.length > 1) {
			const servicesNames = definitions.map(s => s.name.toString());
			throw ERRORS.AMBIGUOUS_SERVICE.create(servicesNames, this.lookup);
		}
		return definitions[0];
	}

	private *findMulti(container: Container): Generator<Definition> {
		yield* this.lookup.find(container);
	}
}
