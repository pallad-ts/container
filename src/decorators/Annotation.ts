import { ensureMetadataAttachedToClass } from "../classServiceMetadata";
import { ClassConstructor } from "../types";

// eslint-disable-next-line @typescript-eslint/naming-convention
export function Annotation(annotation: any): ClassDecorator {
	return (clazz: Function) => {
		ensureMetadataAttachedToClass(clazz as ClassConstructor<any>).annotations.push(annotation);
	};
}
