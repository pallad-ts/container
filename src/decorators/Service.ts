import "reflect-metadata";
import { ClassConstructor, ServiceName } from "../types";
import { ensureMetadataAttachedToClass } from "../classServiceMetadata";
import * as is from "predicates";
import { TypeReference } from "../TypeReference";
import { ERRORS } from "../errors";
import { randomName } from "../utils/randomName";
import { ReferenceArgument } from "../arguments/ReferenceArgument";

export const Service = function (name?: ServiceName) {
	return function (constructor: ClassConstructor<any>) {
		const finalName = name || randomName(constructor.name);

		const metadata = ensureMetadataAttachedToClass(constructor);
		metadata.name = finalName;
		const paramTypes: Array<ClassConstructor<any>> = Reflect.getMetadata(
			"design:paramtypes",
			constructor
		);

		if (is.array(paramTypes) && constructor.length > 0) {
			for (const [index, paramType] of paramTypes.entries()) {
				if (metadata.constructorArguments[index]) {
					continue;
				}
				const ref = TypeReference.createFromClass(paramType);
				if (ref === undefined) {
					throw ERRORS.AUTO_WIRING_FAILED.create(
						`constructor (of ${constructor.name}) argument nr: ${index}`
					);
				}
				metadata.constructorArguments[index] = ReferenceArgument.one.type(ref);
			}
		} else if (constructor.length > 0) {
			throw ERRORS.AUTO_WIRING_NO_METADATA.create();
		}
	};
};
