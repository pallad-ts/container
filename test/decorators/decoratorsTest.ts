import { Service } from "@src/decorators/Service";
import { Inject } from "@src/decorators/Inject";
import { Container } from "@src/Container";
import { Annotation } from "@src/decorators/Annotation";
import { TypeReference } from "@src/TypeReference";
import { extractDefinitionFromClass } from "@src/classServiceMetadata";
import { ReferenceArgument } from "@src/arguments/ReferenceArgument";
import { ERRORS } from "@src/errors";
import "@pallad/errors-dev";

describe("decorators", () => {
	it("simple service", () => {
		@Service("SomeService")
		class Test {}

		const definition = extractDefinitionFromClass(Test);
		expect(definition.name).toEqual("SomeService");

		const result = definition.factory(...definition.arguments);
		expect(result).toBeInstanceOf(Test);
	});

	it("simple service without explicit name", () => {
		@Service()
		class TestFoo {}

		const definition = extractDefinitionFromClass(TestFoo);
		expect(definition.name).toMatch(/^TestFoo.*/);
	});

	it("service with injected args", () => {
		@Service()
		class Foo {
			constructor(@Inject("a") readonly arg1: any) {}
		}

		const definition = extractDefinitionFromClass(Foo);
		expect(definition.name).toMatch(/^Foo.*/);

		expect(definition.arguments).toEqual([ReferenceArgument.one.name("a")]);
	});

	it("Injecting advanced reference", () => {
		const ref = ReferenceArgument.one.annotation((ann): ann is any => true);

		@Service()
		class Foo {
			constructor(@Inject(ref) readonly arg1: any) {}
		}

		const definition = extractDefinitionFromClass(Foo);

		expect(definition.arguments).toEqual([ref]);
	});

	it("injecting type ref", () => {
		class Bar {}

		const ref = TypeReference.createFromClass(Bar)!;

		@Service()
		class Foo {
			constructor(@Inject(ref) readonly prop: Bar) {}
		}

		const definition = extractDefinitionFromClass(Foo);
		expect(definition.arguments).toEqual([ReferenceArgument.one.type(ref)]);
	});

	it("Inject accepts only string or object of ContainerArg class", () => {
		expect(() => {
			@Service()
			class Foo {
				constructor(@Inject([] as any) readonly arg1: any) {}
			}
		}).toThrowErrorMatchingSnapshot();
	});

	it("service with injected args and properties", () => {
		@Service()
		class Foo {
			@Inject("bar")
			bar: any;

			constructor(
				@Inject("a") private arg1: any,
				@Inject("b") private arg2: any
			) {}
		}

		const definition = extractDefinitionFromClass(Foo);
		expect(definition.name).toMatch(/^Foo.*/);

		expect(definition.arguments).toEqual([
			ReferenceArgument.one.name("a"),
			ReferenceArgument.one.name("b"),
			ReferenceArgument.one.name("bar"),
		]);

		const result = definition.factory("a", "b", "bar");

		expect(result).toBeInstanceOf(Foo);

		expect(result.bar).toEqual("bar");

		expect(result.arg1).toEqual("a");

		expect(result.arg2).toEqual("b");
	});

	it("fails if not all constructor arguments have @Inject decorators", () => {
		expect(() => {
			@Service()
			class Foo {
				constructor(
					@Inject("a") readonly arg1: any,
					readonly arg2: unknown
				) {}
			}
		}).toThrowErrorWithCode(ERRORS.AUTO_WIRING_FAILED);
	});

	it("fails it there is an argument, in the middle of other arguments, without @Inject decorator", () => {
		expect(() => {
			@Service()
			class Foo {
				constructor(
					@Inject("a") readonly arg1: any,
					readonly arg2: any,
					@Inject("c") readonly arg3: any
				) {}
			}
		}).toThrowErrorWithCode(ERRORS.AUTO_WIRING_FAILED);
	});

	it("Defining annotation", () => {
		const annotation1 = { name: "test" };
		const annotation2 = { name: "test2" };

		@Annotation(annotation1)
		@Annotation(annotation2)
		@Service()
		class Test {}

		const definition = extractDefinitionFromClass(Test);
		expect(definition.annotations).toEqual([annotation2, annotation1]);
	});

	it("defining annotation when @Service decorator is used first", () => {
		const annotation1 = { name: "test" };
		const annotation2 = { name: "test2" };

		@Service()
		@Annotation(annotation1)
		@Annotation(annotation2)
		class Test {}

		const definition = extractDefinitionFromClass(Test);
		expect(definition.annotations).toEqual([annotation2, annotation1]);
	});

	describe("autowiring", () => {
		let container: Container;

		const NAME = "example";

		class Foo implements FooInterface {
			// tslint:disable-next-line:no-empty
			bonk() {}
		}

		interface FooInterface {
			bonk(): void;
		}

		class Bar {}

		beforeEach(() => {
			container = new Container();
			container.definitionWithConstructor("foo", Foo);
			container.definitionWithConstructor("bar", Bar);
		});

		it("simple", async () => {
			@Service(NAME)
			class Example {
				constructor(readonly foo: Foo) {}
			}

			container.loadDefinitionFromClass(Example);
			const service = await container.resolve<Example>(NAME);

			expect(service).toBeInstanceOf(Example);

			expect(service.foo).toBeInstanceOf(Foo);
		});

		it("with properties", async () => {
			@Service(NAME)
			class Example {
				@Inject()
				bar!: Bar;

				@Inject()
				foo!: Foo;
			}

			container.loadDefinitionFromClass(Example);
			const service = await container.resolve<Example>(NAME);

			expect(service.bar).toBeInstanceOf(Bar);
			expect(service.foo).toBeInstanceOf(Foo);
		});

		it("nothing to autowire", async () => {
			@Service(NAME)
			class Example {}

			container.loadDefinitionFromClass(Example);
			const service = await container.resolve<Example>(NAME);
			expect(service).toBeInstanceOf(Example);
		});

		describe("overwriting an argument", () => {
			it("with type reference", async () => {
				@Service(NAME)
				class Example {
					constructor(@Inject(Bar) readonly foo: Foo) {}
				}

				container.loadDefinitionFromClass(Example);
				const service = await container.resolve<Example>(NAME);

				expect(service.foo).toBeInstanceOf(Bar);
			});

			it("with explicit reference", async () => {
				@Service(NAME)
				class Example {
					constructor(@Inject("bar") readonly foo: Foo) {}
				}

				container.loadDefinitionFromClass(Example);

				const service = await container.resolve<Example>(NAME);
				expect(service.foo).toBeInstanceOf(Bar);
			});

			it("interface reference", async () => {
				@Service(NAME)
				class Example {
					constructor(@Inject(Foo) readonly foo: FooInterface) {}
				}

				container.loadDefinitionFromClass(Example);

				const service = await container.resolve<Example>(NAME);
				expect(service.foo).toBeInstanceOf(Foo);
			});
		});

		describe("fails", () => {
			it("could not autowire interfaces", () => {
				expect(() => {
					@Service(NAME)
					class Example {
						constructor(readonly foo: FooInterface) {}
					}

					container.loadDefinitionFromClass(Example);
				}).toThrowErrorMatchingSnapshot();
			});

			it("could not use @Inject without arg for arguments", () => {
				expect(() => {
					const Inj = Inject as any;

					@Service(NAME)
					class Example {
						constructor(@Inj() readonly arg: any) {}
					}

					container.loadDefinitionFromClass(Example);
				}).toThrowErrorMatchingSnapshot();
			});

			it("could not autowire union types", () => {
				expect(() => {
					@Service()
					class Example {
						constructor(readonly arg: Foo | Bar) {}
					}

					container.loadDefinitionFromClass(Example);
				}).toThrowErrorMatchingSnapshot();
			});

			it("multiple instances of same type", () => {
				container.definitionWithConstructor("another", Foo);

				@Service(NAME)
				class Example {
					constructor(readonly foo: Foo) {}
				}

				container.loadDefinitionFromClass(Example);
				return expect(container.resolve(NAME)).rejects.toThrowErrorWithCode(
					ERRORS.AMBIGUOUS_SERVICE
				);
			});

			it("explicit referencing reserved type", () => {
				expect(() => {
					@Service(NAME)
					class Example {
						constructor(@Inject(Object) readonly foo: Foo) {}
					}

					container.loadDefinitionFromClass(Example);
				}).toThrowErrorMatchingSnapshot();
			});
		});
	});
});
