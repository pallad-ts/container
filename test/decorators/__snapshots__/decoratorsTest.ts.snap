// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorators Inject accepts only string or object of ContainerArg class 1`] = `"@Inject argument must be a string that represents service name, symbol, function or instance of ContainerArg or TypeRef"`;

exports[`decorators autowiring fails could not autowire interfaces 1`] = `"Inferred type for constructor (of Example) argument nr: 0 is an interface, union type or built-in object that is not supported"`;

exports[`decorators autowiring fails could not autowire union types 1`] = `"Inferred type for constructor (of Example) argument nr: 0 is an interface, union type or built-in object that is not supported"`;

exports[`decorators autowiring fails could not use @Inject without arg for arguments 1`] = `"Using @Inject decorator without ref for function arguments is prohibited"`;

exports[`decorators autowiring fails explicit referencing reserved type 1`] = `"Inferred type for constructor (of Example) argument nr: 0 is an interface, union type or built-in object that is not supported"`;
